keps/sig-storage/20200319-selinux-relabeling.md

```diff
@@ -0,0 +1,430 @@
+)
+
+type PodSecurityContext struct {
+    // SELinuxRelabelPolicy ← new field
+    // Defines behavior of changing SELinux labels of the volume before being exposed inside Pod.
+    // Valid values are "OnVolumeMount" and "Always". If not specified, "Always" is used.
+    // "Always" policy recursively changes SELinux labels on all files on all volumes used by the Pod.
+    // "OnVolumeMount" tries to mount volumes used by the Pod with the right context and skip recursive ownership
+    // change.
+    // + optional
+    SELinuxRelabelPolicy *SELinuxRelabelPolicy
+
+    // For context:
+    // fsGroupChangePolicy defines behavior of changing ownership and permission of the volume
+    // before being exposed inside Pod. This field will only apply to
+    // volume types which support fsGroup based ownership(and permissions).
+    // It will have no effect on ephemeral volume types such as: secret, configmaps
+    // and emptydir.
+    // Valid values are "OnRootMismatch" and "Always". If not specified defaults to "Always".
+    // +optional
+    FSGroupChangePolicy *PodFSGroupChangePolicy `json:"fsGroupChangePolicy,omitempty" protobuf:"bytes,9,opt,name=fsGroupChangePolicy"`
+
+    ...
+}
+```
+
+See https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20200120-skip-permission-change.md for similar API for ownership change for fsGroup.
+This KEP should follow API provided for fsGroup closely, however, the implementation is different (`mount` here vs. recursive `chown` in the other KEP).
+
+In order to allow `SELinuxRelabelPolicy: OnVolumeMount` for volumes provided by CSI drivers, kubelet must know if a CSI driver supports SELinux or not.
+
+```go
+// In storage.k8s.io/v1:
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    // supports SELinux specifies that volumes provided by the the CSI driver support
+    // filesystem extended attributes and SELinux. Typically, volumes based on a block
+    // device support SELinux and volumes based on a shared filesystem (NFS, CIFS, CephFS...)
+    // don't.
+    //
+    // When "true", Kubernetes may call NodeStage / NodePublish with "-o context=xyz" mount
+    // option for volumes of a pod with
+    // podSecurityContext.seLinuxRelabelPolicy ="OnVolumeMount" or the container runtime
+    // may change label of all files on the volume to match the Pod requirements.
+    //
+    // When "false", Kubernetes won't pass any special SELinux mount options nor change
+    // SELinux labels on files on the volume.
+    //
+    // When not set, Kubernetes won't pass any special SELinux mount options, but it will
+    // autodetect if a volume supports SELinux or not by checking for "seclabel" mount option
+    // on a NodePublished directory. If "seclabel" is present (i.e. kernel supports SELinux
+    // labels for this volume", container runtime may change label of all files on the volume
+    // to match the Pod requirements.
+    SupportsSELinux *bool;
```

Comment:
Another option is perhaps - to have a field called `StorageIsolationPolicy` or `SecurityPolicy` and define a enum for it with possible value of `selinux` for now. In future this field could be expanded to include other security policies such as apparmour or something else in windows.

Reply:
As I pointed above, the name (and actually also its description) is misleading. The flag is uses to  determine if `mount -o context` on a volume can affect other volumes of the same CSI driver on the same node.

If [nfs-client provisioner](https://github.com/kubernetes-incubator/external-storage/tree/master/nfs-client) had a CSI driver, mounting one of its volume with `-o context=A` sets the context *for all other volumes on the node*, because they come from the same NFS export.

If [nfs provisioner](https://github.com/kubernetes-incubator/external-storage/blob/master/nfs) had a CSI driver, mounting one of its volume with `-o context=A` does not affect the other volumes, because they use separate NFS exports [here I am not 100% sure what ganesha actually does, but I believe there are other NFS servers where it would work].

So the flag is actually about the server and independence of its volumes, not about SELinux or apparmor. It needs a better name.

Reply:
Added more description for `SupportsSELinux` and table with examples.

Reply:
Following my review comment for the fsgroup field, I think it would be better to have explicit enums for the 3 behaviors.

Reply:
I renamed the field to `SELinuxMountSupported` to emphasize that it is only about mounts with `-o context`. It reduces number of cases to handle in kubelet / CSI volume plugin. If the mounted filesystem supports SELinux or not is always autodetected by presence of `seclabel` mount option, as it is already done now.

Therefore the values are `true` or `false`. `nil` means `false`. Do you still want enum? It would be something line `"Supported"` and `"Unsupported"`, which looks odd.